src/poseidon/mod.rs

```diff
@@ -0,0 +1,377 @@
+        }
+        self.state = state;
+    }
+
+    // Absorbs everything in elements, this does not end in an absorbtion.
+    fn absorb_internal(&mut self, mut rate_start_index: usize, elements: &[F]) {
+        let mut remaining_elements = elements;
+
+        loop {
+            // if we can finish in this call
+            if rate_start_index + remaining_elements.len() <= self.rate {
+                for (i, element) in remaining_elements.iter().enumerate() {
+                    self.state[i + rate_start_index] += element;
+                }
+                self.mode = PoseidonSpongeMode::Absorbing {
+                    next_absorb_index: rate_start_index + remaining_elements.len(),
+                };
+
+                return;
+            }
+            // otherwise absorb (rate - rate_start_index) elements
+            let num_elements_absorbed = self.rate - rate_start_index;
+            for (i, element) in remaining_elements
+                .iter()
+                .enumerate()
+                .take(num_elements_absorbed)
+            {
+                self.state[i + rate_start_index] += element;
+            }
+            self.permute();
+            // the input elements got truncated by num elements absorbed
+            remaining_elements = &remaining_elements[num_elements_absorbed..];
+            rate_start_index = 0;
+        }
+    }
+
+    // Squeeze |output| many elements. This does not end in a squeeze
+    fn squeeze_internal(&mut self, mut rate_start_index: usize, output: &mut [F]) {
+        let mut output_remaining = output;
+        loop {
+            // if we can finish in this call
+            if rate_start_index + output_remaining.len() <= self.rate {
+                output_remaining.clone_from_slice(
+                    &self.state[rate_start_index..(output_remaining.len() + rate_start_index)],
+                );
+                self.mode = PoseidonSpongeMode::Squeezing {
+                    next_squeeze_index: rate_start_index + output_remaining.len(),
+                };
+                return;
+            }
+            // otherwise squeeze (rate - rate_start_index) elements
+            let num_elements_squeezed = self.rate - rate_start_index;
+            output_remaining[..num_elements_squeezed].clone_from_slice(
+                &self.state[rate_start_index..(num_elements_squeezed + rate_start_index)],
+            );
+
+            // Unless we are done with squeezing in this call, permute.
+            if output_remaining.len() != self.rate {
+                self.permute();
+            }
+            // Repeat with updated output slices
+            output_remaining = &mut output_remaining[num_elements_squeezed..];
+            rate_start_index = 0;
+        }
+    }
+}
+
+/// Parameters and RNG used
+#[derive(Clone, Debug)]
+pub struct PoseidonParameters<F: PrimeField> {
+    full_rounds: u32,
+    partial_rounds: u32,
+    alpha: u64,
+    mds: Vec<Vec<F>>,
+    ark: Vec<Vec<F>>,
```

Comment:
Are these matrices usually of different sizes, or are they usually fixed size (eg 3x3?). If the latter, we can make them into arrays.

Reply:
sounds good to me. something like `[[F; 3];3]` should work? (please confirm @Will-Lin4 )

Reply:
I had repurposed Poseidon's code from Weikeng's implementation in Marlin for this library, so he may be a more appropriate source to ask. @weikengchen

Reply:
For MDS, yes, it would be a matrix of "t" x "t", where t is the size of the state (rate + capacity). 

For ARK, it is more complicated. The size of ark: Vec<Vec<F>> is "r" x "t", where r is the number of rounds.

Both are related to other parameters in the PoseidonParameters

t = rate and capacity. **Warning: currently the rate and capacity are not in the parameters. We need to put them into the parameters unless we want to support only `let rate = 2;` and `let capacity = 1;`.**

r = self.partial_rounds + self.full_rounds

Do you think that rate and capacity should also be part of the parameters?

Reply:
If rate and capacity are fixed, then we can use arrays instead of vectors. Is there any circumstance where a customizable set of rate and capacity parameters are preferred?

Reply:
@ValarDragon do you have some input on this?